jsonsocket.py

```diff
@@ -1,134 +1,136 @@
+            self.client.close()
+            self.client = None
+        if self.socket:
+            self.socket.close()
+            self.socket = None
 
 
 class Client(object):
-  """
-  A JSON socket client used to communicate with a JSON socket server. All the
-  data is serialized in JSON. How to use it:
-
-  data = {
-    'name': 'Patrick Jane',
-    'age': 45,
-    'children': ['Susie', 'Mike', 'Philip']
-  }
-  client = Client()
-  client.connect(host, port)
-  client.send(data)
-  response = client.recv()
-  # or in one line:
-  response = Client().connect(host, port).send(data).recv()
-  """
-
-  socket = None
-
-  def __del__(self):
-    self.close()
-
-  def connect(self, host, port):
-    self.socket = socket.socket()
-    self.socket.connect((host, port))
-    return self
-
-  def send(self, data):
-    if not self.socket:
-      raise Exception('You have to connect first before sending data')
-    _send(self.socket, data)
-    return self
-
-  def recv(self):
-    if not self.socket:
-      raise Exception('You have to connect first before receiving data')
-    return _recv(self.socket)
-
-  def recv_and_close(self):
-    data = self.recv()
-    self.close()
-    return data
-
-  def close(self):
-    if self.socket:
-      self.socket.close()
-      self.socket = None
-
-
-## helper functions ##
-
+    """
+    A JSON socket client used to communicate with a JSON socket server. All the
+    data is serialized in JSON. How to use it:
+    data = {
+        'name': 'Patrick Jane',
+        'age': 45,
+        'children': ['Susie', 'Mike', 'Philip']
+    }
+    client = Client()
+    client.connect(host, port)
+    client.send(data)
+    response = client.recv()
+    # or in one line:
+    response = Client().connect(host, port).send(data).recv()
+    """
+
+    socket = None
+
+    def __del__(self):
+        self.close()
+
+    def connect(self, host, port):
+        self.socket = socket.socket()
+        self.socket.connect((host, port))
+        return self
+
+    def send(self, data):
+        if not self.socket:
+            raise Exception('You have to connect first before sending data')
+        _send(self.socket, data)
+        return self
+
+    def recv(self):
+        if not self.socket:
+            raise Exception('You have to connect first before receiving data')
+        return _recv(self.socket)
+
+    def recv_and_close(self):
+        data = self.recv()
+        self.close()
+        return data
+
+    def close(self):
+        if self.socket:
+            self.socket.close()
+            self.socket = None
+
+
+# helper functions
 def _send(socket, data):
-  try:
-    serialized = json.dumps(data)
-  except (TypeError, ValueError), e:
-    raise Exception('You can only send JSON-serializable data')
-  # send the length of the serialized data first
-  socket.send('%d\n' % len(serialized))
-  # send the serialized data
-  socket.sendall(serialized)
+    try:
+        serialized = json.dumps(data)
```

Comment:
I think you should encode the data here so that we get the correct length below.